File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/application/services/IExampleService.java
package com.oddly.ddd.application.services;

import com.oddly.ddd.api.dto.v1.requests.CreateExampleRequest;
import com.oddly.ddd.api.dto.v1.requests.UpdateExampleRequest;
import com.oddly.ddd.api.dto.v1.responses.ExampleResponse;

import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Service interface for Example domain.
 * Service interfaces:
 * - Live in /application/services/
 * - Start with I prefix
 * - Implementations go in /application/services/impl/
 * - Orchestrate use-cases, transactions, and policies
 * - Call repositories and domain services
 * - NO business logic (delegate to domain models)
 */
public interface IExampleService {
    /**
     * Creates a new example.
     * Transaction managed by UnitOfWork middleware.
     */
    CompletableFuture<String> createExampleAsync(
        CreateExampleRequest p_request, 
        String p_userId, 
        String p_correlationId
    );

    /**
     * Updates an existing example.
     * Transaction managed by UnitOfWork middleware.
     */
    CompletableFuture<Void> updateExampleAsync(
        String p_id, 
        UpdateExampleRequest p_request, 
        String p_userId
    );

    /**
     * Deletes an example.
     * Transaction managed by UnitOfWork middleware.
     */
    CompletableFuture<Void> deleteExampleAsync(String p_id, String p_userId);

    /**
     * Gets an example by ID.
     * Uses read entity for optimized query.
     */
    CompletableFuture<ExampleResponse> getExampleAsync(String p_id);

    /**
     * Lists all examples with pagination.
     * Uses read entities for optimized queries.
     */
    CompletableFuture<List<ExampleResponse>> listExamplesAsync(int p_skip, int p_take);

    /**
     * Activates an example.
     */
    CompletableFuture<Void> activateExampleAsync(String p_id, String p_userId);

    /**
     * Deactivates an example.
     */
    CompletableFuture<Void> deactivateExampleAsync(String p_id, String p_userId);
}

File: infrastructures/ddd/csharp/src/Application/Services/IExampleService.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using OddlyDdd.Api.Dto.V1.Requests;
using OddlyDdd.Api.Dto.V1.Responses;

namespace OddlyDdd.Application.Services
{
    /// <summary>
    /// Example service interface demonstrating service layer patterns.
    /// Service interfaces:
    /// - Live in /Application/Services/
    /// - Start with I prefix
    /// - Implementations go in /Application/Services/Impl/
    /// - Orchestrate use-cases, transactions, and policies
    /// - Call repositories and domain services
    /// - NO business logic (delegate to domain models)
    /// </summary>
    public interface IExampleService
    {
        /// <summary>
        /// Creates a new example.
        /// Transaction managed by UnitOfWork middleware.
        /// </summary>
        Task<string> CreateExampleAsync(CreateExampleRequest p_request, string p_userId, string p_correlationId);

        /// <summary>
        /// Updates an existing example.
        /// Transaction managed by UnitOfWork middleware.
        /// </summary>
        Task UpdateExampleAsync(string p_id, UpdateExampleRequest p_request, string p_userId);

        /// <summary>
        /// Deletes an example.
        /// Transaction managed by UnitOfWork middleware.
        /// </summary>
        Task DeleteExampleAsync(string p_id, string p_userId);

        /// <summary>
        /// Gets an example by ID.
        /// Uses read entity for optimized query.
        /// </summary>
        Task<ExampleResponse> GetExampleAsync(string p_id);

        /// <summary>
        /// Lists all examples with pagination.
        /// Uses read entities for optimized queries.
        /// </summary>
        Task<IReadOnlyList<ExampleResponse>> ListExamplesAsync(int p_skip, int p_take);

        /// <summary>
        /// Activates an example.
        /// </summary>
        Task ActivateExampleAsync(string p_id, string p_userId);

        /// <summary>
        /// Deactivates an example.
        /// </summary>
        Task DeactivateExampleAsync(string p_id, string p_userId);
    }
}
